phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UncoveredGlobalIndexRegionScanner.java

```diff
@@ -0,0 +1,387 @@
+        }
+        int index = 0;
+        for (Boolean result : resultsAndFutures.getFirst()) {
+            if (result == null) {
+                Throwable cause = ServerUtil.getExceptionFromFailedFuture(
+                        resultsAndFutures.getSecond().get(index));
+                // there was a failure
+                throw new IOException(exceptionMessage, cause);
+            }
+            index++;
+        }
+    }
+
+    private void scanDataTableRows(long startTime)
+            throws IOException {
+        if (indexToDataRowKeyMap.size() == 0) {
+            state = State.READY;
+            return;
+        }
+        TreeSet<byte[]> dataRowKeys = new TreeSet<>(Bytes.BYTES_COMPARATOR);
+        for (byte[] dataRowKey: indexToDataRowKeyMap.values()) {
+            dataRowKeys.add(dataRowKey);
+        }
+        List<Set<byte[]>> setList = IndexRepairRegionScanner.getPerTaskDataRowKeys(dataRowKeys,
+                regionEndKeys, rowCountPerTask);
+        int taskCount = setList.size();
+        TaskBatch<Boolean> tasks = new TaskBatch<>(taskCount);
+        for (int i = 0; i < taskCount; i++) {
+            addTasksForScanningDataTableRowsInParallel(tasks, setList.get(i), startTime);
+        }
+        submitTasks(tasks);
+        if (state == State.SCANNING_DATA_INTERRUPTED) {
+            state = State.SCANNING_DATA;
+        } else {
+            state = State.READY;
+        }
+    }
+
+    /**
+     * A page of index rows are scanned and then their corresponding data table rows are retrieved
+     * from the data table regions in parallel. These data rows are then joined with index rows.
+     * The join is for adding uncovered columns to index rows.
+     *
+     * This implementation conforms to server paging such that if the server side operation takes
+     * more than pageSizeInMs, a dummy result is returned to signal the client that more work
+     * to do on the server side. This is done to prevent RPC timeouts.
+     *
+     * @param result
+     * @return boolean to indicate if there are more rows to scan
+     * @throws IOException
+     */
+    @Override
+    public boolean next(List<Cell> result) throws IOException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
+        boolean hasMore;
+        region.startRegionOperation();
+        try {
+            synchronized (innerScanner) {
+                if (state == State.READY && !indexRowIterator.hasNext()) {
+                    state = State.INITIAL;
+                }
+                if (state == State.INITIAL) {
+                    indexRows = new ArrayList();
+                    dataRows = Maps.newConcurrentMap();
+                    indexToDataRowKeyMap = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
+                    state = State.SCANNING_INDEX;
+                }
+                if (state == State.SCANNING_INDEX) {
+                    do {
+                        List<Cell> row = new ArrayList<Cell>();
```

Comment:
We are missing check for region closing here.

Reply:
@tkhurana, the check for region closing is currently implemented only by the UngroupedAggregateRegionObserver coproc. When the preClose coproc hook is called on UngroupedAggregateRegionObserver by the region server, the coproc sets a flag and the region scanners on this coproc can check this flag to make sure that they do not prolong the region close operation.  This was implemented especially for long running scanners for index rebuild operations. UncoveredGlobalIndexRegionScanner is coproc agnostic and can be initiated from any coproc. So the region closing check can be executed only if all corpoc implements it. Please note that the region closing checks were implemented when the entire index region is scanned and built in one RPC call. This behavior has changed after we introduced server side paging. Please also note that in this PR, we have two levels of paging. One is based on the number of rows and the other based on the time spent within a single RPC call. So, I agree that region closing check is good in general but the need for it much less after server side paging. We need to restructure the server side code in order to implement region closing everywhere. I think that should be a separate improvement.